cpp/ql/test/library-tests/dataflow/dataflow-tests/test.cpp

```diff
@@ -922,4 +922,18 @@ namespace GlobalArrays {
     sink(global_pointer_static_indirect_2); // clean: global_pointer_static_indirect_2 does not have 2 indirections
     indirect_sink(global_pointer_static_indirect_2); // clean: global_pointer_static_indirect_2 does not have 2 indirections
   }
+}
+
+namespace global_variable_conflation_test {
+  int* global_pointer;
+
+  void def() {
+    global_pointer = nullptr;
+    *global_pointer = source();
+  }
+
+  void use() {
+    sink(global_pointer); // clean
+    sink(*global_pointer); // $ ir MISSING: ast
```

Comment:
I can see that these are the correct results, but I don't understand why we were getting it wrong before (or why the fix is correct).

Reply:
Fair point. It took me a while to understand what was going on as well. Consider the current implementation of `isGlobalUse` on `main`:
```ql
private predicate isGlobalUse(
  GlobalLikeVariable v, IRFunction f, int indirection, int indirectionIndex) {
  exists(VariableAddressInstruction vai |
    vai.getEnclosingIRFunction() = f and
    vai.getAstVariable() = v and
    isDef(_, _, _, vai, indirection, indirectionIndex)
  )
}
```
the intuition is that, if there's a definition of the global variable in the enclosing function, then this function should have a "global use" at the exit point of the function (see my [above answer](https://github.com/github/codeql/pull/15282#discussion_r1448748876) for some background information).

The `indirection` tracks whether it's a definition of `global`, or `*global`, or `**global`, etc.

However, consider what happens if there are two definitions in the same function with different indirections level. i.e., the situation I've added in a testcase:
```cpp
int* global_pointer;

void def() {
  global_pointer = nullptr;
  *global_pointer = source();
}
```
in this case, there are two `VariableAddressInstructions` in the body of `def` (i.e., one computing the address of the first assignment, and one computing the address of the second assignment).

The first assignment writes to `global_pointer`, and thus it indirectly also writes to `*global_pointer` (because when you reassign a pointer you also change the value you get when you read the value it points to). The second assignment only writes to `*global_pointer`. So from the point-of-view of dataflow the body of `def` looks like:
```cpp
void def() {
  global_pointer = nullptr;
  *global_pointer = *nullptr; <-- inserted at the dataflow level
  *global_pointer = source();
}
```
we say that the "indirection level" of the first assignment (i.e., `global_pointer = nullptr`) is 0 because it's directly visible at the source code level, and the indirection level of `*global_pointer = *nullptr;` is 1 because it's a definition that occurs because of a definition at indirection level 0.

If `global_pointer` instead was a `int**`, and the following assignment appeared in the source code:
```cpp
global_pointer = nullptr
```
then we'd insert two additional definition, namely `*global_pointer = *nullptr` (as before: with indirection level 1), and `**global_pointer = **nullptr` (with indirection level 2 because it occurs because there's an definition at indirection level 1).

Similarly, the indirection level of the last definition (i.e., `*global_pointer = source();`) is 0 because it appears at the source code level.

So we have three definitions in the `def` function:
1. A definition of `global_pointer` with indirection 0 and indirection level 0
2. A definition of `global_pointer` with indirection 1 and indirection level 1
3. A definition of `global_pointer` with indirection 1 and indirection level 0.

And that's a problem! Because the global use [IPA branch](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L119) is defined by three columns: the variable, the function and an indirection index. And it's expected that the indirection index determine the indirection. But in this case there's two definitions at indirection level 0 with two different indirections.

So that's why we get a conflation problem: because we end up with a "global use" SSA definition that represents multiple indirections.

The fix is to simply get rid of the third definition when computing the set of "global uses". And that's what the `min` is doing.

Reply:
I think I had missed the difference between "indirection" and "indirection level", and also "indirection index" in the source code which I'm still insure about.  Are these terms permanently defined anywhere (e.g. in QLDoc)?

I can see that we want to keep exactly one definition at each indirection, but I haven't quite grasped why we want to keep definitions 1 and 2 in your example above rather than 1 and 3.

Reply:
> I think I had missed the difference between "indirection" and "indirection level", and also "indirection index" in the source code which I'm still insure about. Are these terms permanently defined anywhere (e.g. in QLDoc)?

There are many mentions of indirection levels in the C/C++ dataflow files, but I don't actually think there's a concrete definition of it.

I'll add that in a PR following this one, if that's okay.

Reply:
> I can see that we want to keep exactly one definition at each indirection, but I haven't quite grasped why we want to keep definitions 1 and 2 in your example above rather than 1 and 3.

It's not like we keep any of the definitions. The [TGlobalUse](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L119) branch is just defined in terms of three entities: The global variable, a function `f` that writes to the variable, and an indirection index to make sure there's an SSA use for each indirection of the variable (i.e., there are two global uses for a variable of type `int*`, and three global uses for a variable of type `int**`).

Reply:
To be more concrete then, I haven't worked out why the code isn't:
```
exists(VariableAddressInstruction vai |
    vai.getEnclosingIRFunction() = f and
    vai.getAstVariable() = v and
    isDef(_, _, _, vai, indirection, _)
)
and indirectionIndex = 0
```
... but then I'm currently conflating "indirection level" with "indirection index" so it's no surprise if I'm way off here.

Reply:
Since the `TGlobalUseBranch` looks like:
```ql
TGlobalUse(GlobalLikeVariable v, IRFunction f, int indirectionIndex) {
  isGlobalUse(v, f, _, indirectionIndex)
}
```
the effect of always having `indirectionIndex = 0` would mean that we'd only have one branch for each indirection, which means that we'd be conflating all pointer indirections.

Reply:
After some thought I think I'm happy with this now. :+1: